cpp/src/io/utilities/data_sink.cpp

```diff
@@ -34,7 +36,13 @@ class file_sink : public data_sink {
   explicit file_sink(std::string const& filepath)
   {
     _output_stream.open(filepath, std::ios::out | std::ios::binary | std::ios::trunc);
-    CUDF_EXPECTS(_output_stream.is_open(), "Cannot open output file");
+    if (!_output_stream.is_open()) {
+      auto const dir_path = std::filesystem::path(filepath).parent_path();
+      if (not std::filesystem::exists(dir_path)) {
+        CUDF_FAIL("Cannot open output file; directory does not exist");
+      }
+      CUDF_FAIL("Cannot open output file");
+    }
```

Comment:
would it help to make this code also similar to `cpp/src/io/utilities/file_io_utilities.cpp` ? That seems more detailed.

Reply:
added errno. I originally excluded it because I didn't know if `ofstream::open` sets it.

Reply:
Should we move this into `file_io_utilities`? As it should contain all file IO.

Reply:
TIL `std::strerror`

I had to look up where the allocation happens for the string returned by `std::strerror()`. I'm not convinced it is threadsafe, given that it's allowed to reuse the same allocated space to return strings across threads.

Should we consider switching this to `std::strerror_s()`?

Reply:
great catch! I'll look into `strerror_s`.

Reply:
> Should we move this into file_io_utilities?

I don't think `file_sink` belongs in utilities, but I'll try to reuse this error checking and place the common code in `file_io_utilities`.

Reply:
I should have just put a mutex in here. `strerror_s` portability is horrendous.
Our compiler does not support it, so I have to use `strerror_r`. Aaand, this one is problematic as well, see https://www.club.cc.cmu.edu/~cmccabe/blog_strerror.html